# Test analysis with a new dataset GSE30528Â´

library("affy")
library("GEOquery")
library("tidyverse")

# The analysis begins with the .CEL files generated by affymetrix RNA image analysis
# which contain the raw information extracted from the probes.They need to undergo normalization first.

getGEOSuppFiles('GSE30528') # get the .CEL files from GEO

untar("GSE30528/GSE30528_RAW.tar", exdir = 'data_RAW_2/') # extract the raw files

raw.data = ReadAffy(celfile.path='data_RAW_2/') # read the .cel files

# ReadAffy is an AffyBatch object that stores the data from the .cel files for further analysis

norm.data = rma(raw.data) # perform sample normalization

norm.expr = exprs(norm.data) # fetch the normalized data in a matrix form

df.normexp = as.data.frame(norm.expr) # transform the matrix into a df

gse = getGEO('GSE30528', GSEMatrix = TRUE) # map the probes ID to gene symbols

feature.data = gse$GSE30528_series_matrix.txt.gz@featureData@data # indexing of the gse object to fetch the genes ID for different classification (feature data -> data)

feature.data = feature.data[,c(1,11)] # recover the affymetrix ID and gene symbol

samples <- colnames(df.normexp)

heatmap(cor(df.normexp))

install.packages("hclust")
library('hclust')

#samples = c("GSM15965.CEL.gz", "GSM15966.CEL.gz", "GSM15967.CEL.gz", "GSM15968.CEL.gz", "GSM15969.CEL.gz", "GSM15970.CEL.gz")

#write.table(df.normexp, file = "matrix_GSE30528", sep = "\t", row.names = T)

feature.data$ID[feature.data$`Gene Symbol` %in% c('CXCL9', 'CCL5', 'CCL19'),]

#____________________________Differential gene expression analysis____________________________

library("limma")

design <- model.matrix(~0 + factor(c(rep("Replicate", 9), factor(c(rep("Control", 22-9)))))) # Create a design matrix

# Assign column names to the design matrix
colnames(design) <- c("Controls","Replicates")
rownames(design) <- samples

contrasts = makeContrasts(Replicates - Controls, levels = design) # Define contrasts to compare each replicate with its respective control

fit <- lmFit(df.normexp[samples], design, ) # fit the data in a linear model. This creates a MArrayLM object that is read further down

fit <- contrasts.fit(fit, contrasts = contrasts) # perform comparison based on predetermined groups

fit <- eBayes(fit) # evaluate statistics of each signal

results.fit <- decideTests(fit)

vennDiagram(results.fit)

exp.result <- topTable(fit, adjust.method = 'BH',number=Inf) %>% # show results with adjusted p_val
  rownames_to_column(var = 'ID') %>%
  inner_join(.,feature.data, by = 'ID')%>%
  subset(logFC >= 1 & adj.P.Val < 0.05)
  


write(upreg.result$Gene.Symbol,file='upreg',sep="\t")
write(exp.result.collapsed$Gene.Symbol,file='universe',sep='\t')

#write.table(exp.result, file = "Diff_analysis_GSE30528", sep = "\t", row.names = T)

exp.result.collapsed <- read.table('collapsed_DEGA_GSE30528.txt',header = T,sep='\t')

# All DEGs

exp.result.filtered <- exp.result.collapsed[exp.result.collapsed$adj.P.Val < 0.05 & abs(exp.result.collapsed$logFC) >= 1,] # filter for p_val nad logFC
diff.genes <- exp.result.filtered$Gene.Symbol # list with differentially expressed genes

# Up-regulated genes
upreg.result <- exp.result.collapsed %>%
  subset(logFC >= 1 & adj.P.Val < 0.05)

# Down-regulated genes
downreg.result <- exp.result.collapsed %>%
  subset(logFC <= -1 & adj.P.Val < 0.05)

# Save gene entries in a list

upreg.genes <- upreg.result$`Gene.Symbol`

downreg.genes <- downreg.result$`Gene.Symbol`

gene.list <- exp.result.filtered$`Gene.Symbol`

# check probes with no related genes
all.genes <- c()
for(entry in exp.result$`Gene Symbol`){
  if(entry == ''){
    next
  }
  else{
    all.genes <- c(all.genes,entry)
  }
}

# volcano plot to check the generalities of the results

library(ggplot2)

ggplot(exp.result.collapsed, aes(x = logFC, y = -log10(adj.P.Val))) + #ggplot initializes the ggplot object and serves as a resevoir for the df obj and the vectors in it that will serve as variables
  geom_point(aes(color = ifelse(logFC >= 1 & adj.P.Val < 0.05, "red", 
                                ifelse(logFC <= -1 & adj.P.Val < 0.05, "blue", "gray")), alpha = 0.7))  +
  scale_color_identity() +
  labs(x = "Log Fold Change", y = "-log10(p-value)", title = "Volcano Plot") +
  theme_minimal() +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "gray") +
  geom_vline(xintercept = 1, linetype = "solid", color = "gray") +
  geom_vline(xintercept = -1, linetype = "solid", color = "gray")

ggplot(exp.result.collapsed, aes(x = logFC, y = -log10(adj.P.Val))) +
  geom_point(
    aes(color = ifelse(logFC >= 1 & adj.P.Val < 0.05, "red", 
                       ifelse(logFC <= -1 & adj.P.Val < 0.05, "blue", "gray")), alpha = 0.7)
  ) +
  geom_text(
    aes(label = `Gene.Symbol`),  # Assuming 'gene_symbol' is the column with gene labels
    position = position_nudge(y = 0.2),  # Adjust the position of labels for better visibility
    size = 3,  # Adjust the size of labels
    check_overlap = TRUE  # Allow labels to overlap if necessary
  ) +
  scale_color_identity() +
  labs(x = "Log Fold Change", y = "-log10(p-value)", title = "Volcano Plot") +
  theme_minimal() +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "gray") +
  geom_vline(xintercept = 1, linetype = "solid", color = "gray") +
  geom_vline(xintercept = -1, linetype = "solid", color = "gray")

#____________________________Functional Enrichment____________________________

library(clusterProfiler)
library(org.Hs.eg.db)

# Retrieve the universe to be used in the over-representation step

universe <- exp.result.collapsed$`Gene.Symbol` # Extract all genes identified by the affymetrix analysis

# GO over-representation with all genes_________________________________________________________________________


enric.bp <- enrichGO(diff.genes, # perform GO enrichment analysis (over-representation analysis)
                     OrgDb=org.Hs.eg.db, 
                     keyType="SYMBOL",
                     ont="All",
                     pvalueCutoff = 0.05, # cutoff for adjusted p_val
                     pAdjustMethod = "BH", 
                     universe=universe, 
                     minGSSize = 20, 
                     maxGSSize=300, 
                     readable = T)

library(DOSE)

enric.bp@result$FoldEnrichment=parse_ratio(enric.bp@result$GeneRatio)/parse_ratio(enric.bp@result$BgRatio) # add a column with the fold enrichment, defined as the relative amount of time a gene appears in a GO term in the gene list when compared to the universe

enric.bp.df <- as.data.frame(enric.bp)

# Semantic similarity analysis to correct for similar terms

enricbp.simp <- simplify(enric.bp,cutoff = 0.3) # remove redundant terms based on their level of proximity in the GO hierarchy

enric.simp.df <- as.data.frame(enricbp.simp)
#write.table(enric.simp.df, file = "GOFunctionalEnrichment_Simp.txt", sep = "\t", row.names = T)

dotplot(enricbp.simp, x="FoldEnrichment",color="p.adjust",size="Count",showCategory=20, orderBy="Count",title="All Genes") # plot the results as a dotplot

# GO over-representation with up-regulated genes_________________________________________________________________________

up.enric.bp <- enrichGO(upreg.genes, 
                     OrgDb=org.Hs.eg.db, 
                     keyType="SYMBOL",
                     ont="All",
                     pvalueCutoff = 0.05,
                     pAdjustMethod = "BH", 
                     universe=universe, 
                     minGSSize = 20, 
                     maxGSSize=300, 
                     readable = T)

up.enric.bp@result$FoldEnrichment=parse_ratio(up.enric.bp@result$GeneRatio)/parse_ratio(up.enric.bp@result$BgRatio) # add a column with the fold enrichment, defined as the relative amount of time a gene appears in a GO term in the gene list when compared to the universe

up.enric.bp.df <- as.data.frame(up.enric.bp)
up.todrop <- up.enric.bp.df$ID[up.enric.bp.df$Count < 5]
up.enric.bp <- dropGO(up.enric.bp,term=up.todrop)
# Semantic similarity analysis to correct for similar terms

up.enricbp.simp <- simplify(up.enric.bp,cutoff = 0.5) # remove redundant terms based on their level of proximity in the GO hierarchy

up.enric.simp.df <- as.data.frame(up.enricbp.simp)
up.enricbp.simp <- dropGO(up.enricbp.simp,term=c('GO:0060205','GO:0031983'))

up.enric.simp.df <- as.data.frame(up.enricbp.simp)
#write.table(up.enric.simp.df, file = "UPREG_GOFunctionalEnrichment_Simp.txt", sep = "\t", row.names = T)

dotplot(
  up.enricbp.simp,
  x = "FoldEnrichment",
  color="p.adjust",
  size="Count",
  orderBy='Count',
  showCategory = 5,
  split='ONTOLOGY',
  font.size=16)

#dotplot(up.enricbp.simp, x="FoldEnrichment",color="p.adjust",size="Count",showCategory=20, orderBy="Count",title="Up-Regulated Genes") # plot the results as a dotplot

# GO over-representation with up-regulated genes_________________________________________________________________________

dw.enric.bp <- enrichGO(downreg.genes, 
                        OrgDb=org.Hs.eg.db, 
                        keyType="SYMBOL",
                        ont="All",
                        pvalueCutoff = 0.05,
                        pAdjustMethod = "BH", 
                        universe=universe, 
                        minGSSize = 20, 
                        maxGSSize=300, 
                        readable = T)

dw.enric.bp@result$FoldEnrichment=parse_ratio(dw.enric.bp@result$GeneRatio)/parse_ratio(dw.enric.bp@result$BgRatio) # add a column with the fold enrichment, defined as the relative amount of time a gene appears in a GO term in the gene list when compared to the universe

dw.enric.bp.df <- as.data.frame(dw.enric.bp)

# Semantic similarity analysis to correct for similar terms

dw.enricbp.simp <- simplify(dw.enric.bp,cutoff = 0.4) # remove redundant terms based on their level of proximity in the GO hierarchy

dw.enric.simp.df <- as.data.frame(dw.enricbp.simp)
#write.table(dw.enric.simp.df, file = "DOWNREG_GOFunctionalEnrichment_Simp.txt", sep = "\t", row.names = T)

dotplot(
  dw.enricbp.simp,
  x = "FoldEnrichment",
  color="p.adjust",
  size="Count",
  orderBy="Count",
  showCategory = 5,
  split='ONTOLOGY')

dotplot(up.enricbp.simp, x="FoldEnrichment",color="p.adjust",size="Count",showCategory=20, orderBy="Count",title="Down-Regulated Genes") # plot the results as a dotplot

# KEG pathway enrichment

# All genes:

entrezID <- gse$GSE30528_series_matrix.txt.gz@featureData@data[c(11,12)]

symbol_to_ID <- function(genelist, df){
  gene.listfinal <- c()
  for(gene in genelist){
    entry <- df$ENTREZ_GENE_ID[df$`Gene Symbol` == gene]
    gene.listfinal <- c(gene.listfinal,entry)
  }
  return(gene.listfinal <- unique(gene.listfinal))
}

universeEntrez <- symbol_to_ID(universe, entrezID)

gene.listEntrez <- symbol_to_ID(diff.genes, entrezID)

kk <- enrichKEGG(gene = gene.listEntrez,
                 organism = 'hsa',
                 pAdjustMethod = "BH",
                 universe = universeEntrez,
                 pvalueCutoff = 0.1,
                 minGSSize = 5,
                 maxGSSize = 500)

kk@result$FoldEnrichment=parse_ratio(kk@result$GeneRatio)/parse_ratio(kk@result$BgRatio)

kkdf=as.data.frame(kk)

#write.table(up.kkdf, file = "UPREG_GOKEGGEnrichment.txt", sep = "\t", row.names = T)

barplot(kk)

# Up-regulated genes:

up.gene.listEntrez <- symbol_to_ID(upreg.genes, entrezID)

up.kk <- enrichKEGG(gene = up.gene.listEntrez,
                 organism = 'hsa',
                 pAdjustMethod = "BH",
                 universe = universeEntrez,
                 pvalueCutoff = 0.05,
                 minGSSize = 50,
                 maxGSSize = 300)

up.kk@result$FoldEnrichment=parse_ratio(up.kk@result$GeneRatio)/parse_ratio(up.kk@result$BgRatio)

up.kkdf=as.data.frame(up.kk)

browseKEGG(up.kk, 'hsa04015')


#write.table(dw.kkdf, file = "DOWNREG_GOKEGGEnrichment.txt", sep = "\t", row.names = T)

barplot(up.kk,
        width = 0.4,
        font.size=16)
dotplot(
  up.kk,
  x = "FoldEnrichment",
  color="p.adjust",
  size="Count",
  orderBy="Count",
  showCategory = 7,)

# Down-regulated genes:

dw.gene.listEntrez <- symbol_to_ID(downreg.genes, entrezID)

dw.kk <- enrichKEGG(gene = dw.gene.listEntrez,
                    organism = 'hsa',
                    pAdjustMethod = "BH",
                    universe = universeEntrez,
                    pvalueCutoff = 0.05,
                    minGSSize = 10,
                    maxGSSize = 300)

dw.kk@result$FoldEnrichment=parse_ratio(dw.kk@result$GeneRatio)/parse_ratio(dw.kk@result$BgRatio)

dw.kkdf=as.data.frame(dw.kk)

#write.table(kkdf, file = "GOKEGGEnrichment.txt", sep = "\t", row.names = T)

barplot(dw.kk,
        width = 0.4,
        font.size=16)

dotplot(
  dw.kk,
  x = "FoldEnrichment",
  color="p.adjust",
  size="Count",
  orderBy="Count",
  showCategory = 7,)

#____________________________GSEA____________________________

gsea.df <- exp.result.collapsed[c('logFC', 'Gene.Symbol')]

geneList <- gsea.df[,1]
names(geneList) <- as.character(gsea.df[,2])
geneList = sort(geneList, decreasing = TRUE)

gsea.result <- gseGO(geneList = geneList,
                OrgDb=org.Hs.eg.db,
                keyType = 'SYMBOL',
                ont = "All",  # Biological Process, you can choose "CC" for Cellular Component or "MF" for Molecular Function
                pAdjustMethod = 'BH',  # Number of permutations for statistical testing
                minGSSize = 10,  # Minimum size of the gene set
                maxGSSize = 500,  # Maximum size of the gene set
                pvalueCutoff = 0.1)  # P-value cutoff for significance

gsea.result <- simplify(gsea.result, # semantic similarity analysis
                        cutoff = 0.3)

gsea.result.df <- as.data.frame(gsea.result)

dotplot(gsea.result, x="enrichmentScore",color="p.adjust", size="setSize",showCategory=20, orderBy="setSize", title = 'GSEA BP')

