{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'pandas'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp\\ipykernel_46416\\3213494386.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[1;32mimport\u001b[0m \u001b[0mpandas\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      2\u001b[0m \u001b[0mget_ipython\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mrun_line_magic\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'matplotlib'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'inline'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[0mmatplotlib\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mpyplot\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0mplt\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mseaborn\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0msns\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mgeneddit\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0mge\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mModuleNotFoundError\u001b[0m: No module named 'pandas'"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "%matplotlib inline\n",
    "from matplotlib import pyplot as plt\n",
    "import seaborn as sns\n",
    "import geneddit as ge\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readFASTA(filename):\n",
    "    \"\"\"This function reads a FASTA format file and\n",
    "    returns a pair of strings\n",
    "    with the header and the sequence\n",
    "    \"\"\"\n",
    "    with open(filename) as a:\n",
    "        lines = [line.strip() for line in a] #Get every line in the file, remove the blank space (\\n) and add to the list\n",
    "    lines = [line for line in lines if len(line) > 0] #Remove empty elements on the previous list\n",
    "    \n",
    "    if lines[0].startswith('>'): #Conditional for two possible FASTA types. If the first line is not the protein or DNA line, then separate it\n",
    "        return lines[0], ''.join(lines[1:])\n",
    "    else: #If not, just join all the lines\n",
    "        return '', ''.join(lines)\n",
    "\n",
    "# A função 'open' lê cada linha como um elemento individual do ficheiro.\n",
    "# É quase como se cada linha fosse um elemento da lista gerada com a função.\n",
    "# A função retorna uma string de acordo com as operações mencionadas.\n",
    "\n",
    "basesDNA = 'ATGC'\n",
    "basesRNA = 'AUGC'\n",
    "\n",
    "aa_residues   = \"ACDEFGHIKLMNPQRSTVWY\"\n",
    "\n",
    "complementDNA = { 'A':'T', 'T':'A', 'G':'C', 'C':'G'}\n",
    "complementRNA = { 'A':'U', 'T':'A', 'G':'C', 'C':'G'}\n",
    "\n",
    "gencode = {'UUU': 'F', 'UUC': 'F', 'UUA': 'L', 'UUG': 'L', 'UCU': 'S',\n",
    "     'UCC': 'S', 'UCA': 'S', 'UCG': 'S', 'UAU': 'Y', 'UAC': 'Y',\n",
    "     'UGU': 'C', 'UGC': 'C', 'UGG': 'W', 'CUU': 'L', 'CUC': 'L',\n",
    "     'CUA': 'L', 'CUG': 'L', 'CCU': 'P', 'CCC': 'P', 'CCA': 'P',\n",
    "     'CCG': 'P', 'CAU': 'H', 'CAC': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
    "     'CGU': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R', 'AUU': 'I',\n",
    "     'AUC': 'I', 'AUA': 'I', 'AUG': 'M', 'ACU': 'T', 'ACC': 'T',\n",
    "     'ACA': 'T', 'ACG': 'T', 'AAU': 'N', 'AAC': 'N', 'AAA': 'K',\n",
    "     'AAG': 'K', 'AGU': 'S', 'AGC': 'S', 'AGA': 'R', 'AGG': 'R',\n",
    "     'GUU': 'V', 'GUC': 'V', 'GUA': 'V', 'GUG': 'V', 'GCU': 'A',\n",
    "     'GCC': 'A', 'GCA': 'A', 'GCG': 'A', 'GAU': 'D', 'GAC': 'D',\n",
    "     'GAA': 'E', 'GAG': 'E', 'GGU': 'G', 'GGC': 'G', 'GGA': 'G',\n",
    "     'GGG': 'G', 'UAA': 'STOP', 'UAG': 'STOP', 'UGA': 'STOP'}\n",
    "\n",
    "trans123 = {'A': 'Ala', 'C': 'Cys', 'E': 'Glu', 'D': 'Asp', 'G': 'Gly',\n",
    "            'F': 'Phe', 'I': 'Ile', 'H': 'His', 'K': 'Lys', 'M': 'Met',\n",
    "            'L': 'Leu', 'N': 'Asn', 'Q': 'Gln', 'P': 'Pro', 'S': 'Ser',\n",
    "            'R': 'Arg', 'T': 'Thr', 'W': 'Trp', 'V': 'Val', 'Y': 'Tyr'}\n",
    "\n",
    "aa_masses = {'A': 71.0788, \n",
    "             'C': 103.1448, \n",
    "             'E': 129.1155, \n",
    "             'D': 115.0886, \n",
    "             'G': 57.0519, \n",
    "             'F': 147.1766, \n",
    "             'I': 113.1594, \n",
    "             'H': 137.1411, \n",
    "             'K': 128.1741, \n",
    "             'M': 131.1926, \n",
    "             'L': 113.1594, \n",
    "             'N': 114.1038, \n",
    "             'Q': 128.1307, \n",
    "             'P': 97.11667, \n",
    "             'S': 87.0782, \n",
    "             'R': 156.1875, \n",
    "             'T': 101.1051, \n",
    "             'W': 186.2132, \n",
    "             'V': 99.1326, \n",
    "             'Y': 163.176\n",
    "             }\n",
    "\n",
    "aa_classes = { 'small'       : 'PCAGVTDSN',\n",
    "    'tiny'        : 'AGCS',\n",
    "    'aliphatic'   : 'ILV',\n",
    "    'aromatic'    : 'FYWH',\n",
    "    'positive'    : 'KHR',\n",
    "    'negative'    : 'DE',\n",
    "    'charged'     : 'KHRDE',\n",
    "    'hydrophobic' : 'CAGIVLTMHYWF',\n",
    "    'polar'       : \"CSTNDYWHKQRDE\",\n",
    "    'proline'     : 'P'}\n",
    "\n",
    "\n",
    "def complement(seq):\n",
    "    complementair = []\n",
    "    for n in seq:\n",
    "        complementair.append(complementDNA[n])\n",
    "    comp = ''.join(complementair)\n",
    "    return comp\n",
    "\n",
    "def prot_noncod(seq): #From the noncoding chain it produces the associated protein\n",
    "    comp = complement(seq)\n",
    "    mRNA = []\n",
    "    for n in comp:\n",
    "        mRNA.append(complementRNA[n])\n",
    "    mRNA = ''.join(mRNA)\n",
    "\n",
    "    protein = []\n",
    "    cod = [mRNA[i] + mRNA[i+1] + mRNA[i+2] for i in range(0, len(mRNA), 3)]\n",
    "    for n in cod:\n",
    "        protein.append(gencode[n])\n",
    "    protein = '-'.join(protein)\n",
    "    return protein\n",
    "\n",
    "def prot_cod(seq): #From the noncoding chain it produces the associated protein\n",
    "    mRNA = []\n",
    "    for n in seq:\n",
    "        mRNA.append(complementRNA[n])\n",
    "    mRNA = ''.join(mRNA)\n",
    "\n",
    "    protein = []\n",
    "    cod = [mRNA[i] + mRNA[i+1] + mRNA[i+2] for i in range(0, len(mRNA), 3)]\n",
    "    for n in cod:\n",
    "        protein.append(gencode[n])\n",
    "    protein = '-'.join(protein)\n",
    "    return protein\n",
    "    \n",
    "def BLAST(seq, seq1): #For a sequence (seq), this function analyses if another, smaller sequence (seq1) is contained inside the first one\n",
    "    complementair = []\n",
    "    for n in seq:\n",
    "        complementair.append(complementDNA[n])\n",
    "    comp = ''.join(complementair)\n",
    "\n",
    "    anneal = str(seq1)\n",
    "\n",
    "    if anneal in seq:\n",
    "        return print(f'{anneal} appears in the requested sequence')\n",
    "    elif anneal in comp:\n",
    "        return print(f'{anneal} appears in the sequence complementary to the requested one')\n",
    "    else:\n",
    "        return print(f'{anneal} does not appear in any sequence') \n",
    "\n",
    "def gene_size(gene, primerfw, primerrv):\n",
    "    if primerfw in gene:\n",
    "        a = gene.split(primerfw)[1]\n",
    "        a = primerfw + a\n",
    "        a = a.split(complement(primerrv))[0]\n",
    "        a = a + complement(primerrv)\n",
    "        size = len(a)\n",
    "    return print(f'The requested amplicon has {size} bp')\n",
    "\n",
    "def amplicon(gene, primerfw, primerrv):\n",
    "    try:\n",
    "        if primerfw in gene:\n",
    "            a = gene.split(primerfw)[1]\n",
    "            a = primerfw + a\n",
    "            a = a.split(complement(primerrv))[0]\n",
    "            a = a + complement(primerrv)\n",
    "            return a\n",
    "    except:\n",
    "        print('Foward primer not found')\n",
    "    \n",
    "def reverse(x):\n",
    "    return x[::-1]\n",
    "\n",
    "def mass(seq):\n",
    "    mass_list = []\n",
    "    for aa in seq:\n",
    "        mass = aa_masses[aa]\n",
    "        mass_list.append(mass)\n",
    "    return sum(mass_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene analyser functions' test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCAGTTCCTAGGTCGTTCCT appears in the sequence complementary to the requested one\n",
      "AAGGGGCTTTACGATGGAGT appears in the requested sequence\n",
      "CGATCAGGCGAAAATAGC appears in the requested sequence\n",
      "TACGCGGTGTTTTTTAGTTATA appears in the sequence complementary to the requested one\n",
      "The requested amplicon has 336 bp\n",
      "The requested amplicon has 489 bp\n",
      "336\n"
     ]
    }
   ],
   "source": [
    "import geneddit as ge\n",
    "gene = readFASTA('GLO1_flanking.fsa')[1]\n",
    "gene2 = readFASTA('VPS13_flanking.fsa')[1]\n",
    "\n",
    "#print(gene)\n",
    "#GLO1:\n",
    "primerfw = 'AAGGGGCTTTACGATGGAGT' #5' - 3'\n",
    "primerrv = 'TCAGTTCCTAGGTCGTTCCT' #3' - 5'\n",
    "\n",
    "#VPS13:\n",
    "primerfw1 = 'CGATCAGGCGAAAATAGC'\n",
    "primerrv1 = 'TACGCGGTGTTTTTTAGTTATA'\n",
    "\n",
    "BLAST(gene, primerrv)\n",
    "BLAST(gene, primerfw)\n",
    "BLAST(gene2, primerfw1)\n",
    "BLAST(gene2, primerrv1)\n",
    "gene_size(gene, primerfw, primerrv)\n",
    "gene_size(gene2, primerfw1, primerrv1)\n",
    "amplicon = ge.amplicon(gene, primerfw, primerrv)\n",
    "print(len(amplicon))\n",
    "\n",
    "#print(prot_noncod(gene))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Altering a protein sequence for structure homology modeling "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GAMASSVLVTQEPEIELPREPRPNEECLQILGNAEKGAKFLSDAEIIQLVNAKHIPAYKLETLIETHERGVSIRRQLLSKKLSEPSSLQYLPYRDYNYSLVMGACCENVIGYMPIPVGVAGPLCLDEKEFQVPMATTEGCLVASTNRGCRAIGLGGGASSRVLADGMTRGPVVRLPRACDSAEVKAWLETSEGFAVIKEAFDSTSRFARLQKLHTSIAGRNLYIRFQSRSGDAMGMNMISKGTEKALSKLHEYFPEMQILAVSGNYCTDKKPAAINWIEGRGKSVVCEAVIPAKVVREVLKTTTEAMIEVNINKNLVGSAMAGSIGGYNAHAANIVTAIYIACGQDAAQNVGSSNCITLMEASGPTNEDLYISCTMPSIEIGTVGGGTNLLPQQACLQMLGVQGACKDNPGENARQLARIVCGTVMAGELSLMAALAAGHLVKSHMIHNRSKINLQDLQGACTKKTA\n",
      "GAMASSVLVTQEPEIELPREPRPNEECLQILGNAEKGAKFLSDAEIIQLVNAKHIPAYKLETLIETHERGVSIRRQLLSKKLSEPSSLQYLPYRDYNYSLVMGACCENVIGYMPIPVGVAGPLCLDEKEFQVPMATTEGCLVASTNRGCRAIGLGGGASSRVLADGMTAGPVVRLPRACDSAEVKAWLETSEGFAVIKEAFDSTSRFARLQKLHTSIAGRNLYIRFQSRSGDAMGMNMISKGTEKALSKLHEYFPEMQILAVSGNYCTDKAPAAINWIEGRGKSVVCEAVIPAKVVREVLKTTTEAMIEVNINDNLVGSAMAGSIGGYNAEAANIVTAIYIACGQDAAQNVGSSNCITLMEASGPTNEDLYISCTMPSIEIGTVGGGTNLLPQQACLQMLGVQGACKDNPGENARQLARIVCGTVMAGELSLMAALAAGHLVKSHMIHNRSKINLQDLQGACTKKTA\n",
      "590 R -> A\n",
      "692 K -> A\n",
      "735 K -> D\n",
      "752 H -> E\n",
      "{'G': 42, 'A': 50, 'M': 17, 'S': 33, 'V': 31, 'L': 42, 'T': 24, 'Q': 18, 'E': 33, 'P': 20, 'I': 31, 'R': 22, 'N': 23, 'C': 16, 'K': 24, 'F': 7, 'D': 12, 'H': 8, 'Y': 12, 'W': 2}\n",
      "{'G': 42, 'A': 52, 'M': 17, 'S': 33, 'V': 31, 'L': 42, 'T': 24, 'Q': 18, 'E': 34, 'P': 20, 'I': 31, 'R': 21, 'N': 23, 'C': 16, 'K': 22, 'F': 7, 'D': 13, 'H': 7, 'Y': 12, 'W': 2}\n"
     ]
    }
   ],
   "source": [
    "HMGcoa = 'GAMASSVLVTQEPEIELPREPRPNEECLQILGNAEKGAKFLSDAEIIQLVNAKHIPAYKLETLIETHERGVSIRRQLLSKKLSEPSSLQYLPYRDYNYSLVMGACCENVIGYMPIPVGVAGPLCLDEKEFQVPMATTEGCLVASTNRGCRAIGLGGGASSRVLADGMTRGPVVRLPRACDSAEVKAWLETSEGFAVIKEAFDSTSRFARLQKLHTSIAGRNLYIRFQSRSGDAMGMNMISKGTEKALSKLHEYFPEMQILAVSGNYCTDKKPAAINWIEGRGKSVVCEAVIPAKVVREVLKTTTEAMIEVNINKNLVGSAMAGSIGGYNAHAANIVTAIYIACGQDAAQNVGSSNCITLMEASGPTNEDLYISCTMPSIEIGTVGGGTNLLPQQACLQMLGVQGACKDNPGENARQLARIVCGTVMAGELSLMAALAAGHLVKSHMIHNRSKINLQDLQGACTKKTA'\n",
    "aminoacids = []\n",
    "for aa in HMGcoa:\n",
    "    aminoacids.append(aa)\n",
    "#print(len(aminoacids))\n",
    "index = range(422,889)\n",
    "#print(len(index))\n",
    "df = pd.DataFrame(aminoacids, index, ['aminoacids'])\n",
    "\n",
    "# Insert mutations on the protein sequence\n",
    "df.loc[692] = 'A'\n",
    "df.loc[590] = 'A'\n",
    "df.loc[735] = 'D'\n",
    "df.loc[752] = 'E'\n",
    "\n",
    "mutated_aa = df['aminoacids'].tolist()\n",
    "mutated_HMGcoa = ''.join(mutated_aa)\n",
    "print(HMGcoa)\n",
    "print(mutated_HMGcoa)\n",
    "\n",
    "# Confirming mutation occurred at the right spot\n",
    "for i in range(0,len(HMGcoa)):\n",
    "    if HMGcoa[i] == mutated_HMGcoa[i]:\n",
    "        continue\n",
    "    else: \n",
    "        print(i+422,HMGcoa[i],'->',mutated_HMGcoa[i])\n",
    "\n",
    "count = dict()\n",
    "for aminoacid in HMGcoa:\n",
    "    if aminoacid in count:\n",
    "        count[aminoacid] = count[aminoacid] + 1\n",
    "    else:\n",
    "        count[aminoacid] = 1\n",
    "print(count)\n",
    "count2 = dict()\n",
    "for aminoacid in mutated_HMGcoa:\n",
    "    if aminoacid in count2:\n",
    "        count2[aminoacid] = count2[aminoacid] + 1\n",
    "    else:\n",
    "        count2[aminoacid] = 1\n",
    "print(count2)\n",
    "\n",
    "#pd.DataFrame(data=count,index=['count']).transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MAEDDPYLGRPEQMFHLDPSLTHTIFNPEVFQPQMALPTDGPYLQILEQPKQRGFRFRYVCEGPSHGGLPGASSEKNKKSYPQVKICNYVGPAKVIVQLVTNGKNIHLHAHSLVGKHCEDGICTVTAGPKDMVVGFANLGILHVTKKKVFETLEARMTEACIRGYNPGLLVHPDLAYLQAEGGGDRQLGDREKELIRQAALQQTKEMDLSVVRLMFTAFLPDSTGSFTRRLEPVVSDAIYDSKAPNASNLKIVRMDRTAGCVTGGEEIYLLCDKVQKDDIQIRFYEEEENGGVWEGFGDFSPTDVHRQFAIVFKTPKYKDINITKPASVFVQLRRKSDLETSEPKPFLYYPEIKDKEEVQRKRQKLMPNFSDSFGGGSGAGAGGGGMFGSGGGGGGTGSTGPGYSFPHYGFPTYGGITFHPGTTKSNAGMKHGTMDTESKKDPEGCDKSDDKNTVNLFGKVIETTEQDQEPSEATVGNGEVTLTYATGTKEESAGVQDNLFLEKAMQLAKRHANALFDYAVTGDVKMLLAVQRHLTAVQDENGDSVLHLAIIHLHSQLVRDLLEVTSGLISDDIINMRNDLYQTPLHLAVITKQEDVVEDLLRAGADLSLLDRLGNSVLHLAAKEGHDKVLSILLKHKKAALLLDHPNGDGLNAIHLAMMSNSLPCLLLLVAAGADVNAQEQKSGRTALHLAVEHDNISLAGCLLLEGDAHVDSTTYDGTTPLHIAAGRGSTRLAALLKAAGADPLVENFEPLYDLDDSWENAGEDEGVVPGTTPLDMATSWQVFDILNGKPYEPEFTSDDLLAQGDMKQLAEDVKLQLYKLLEIPDPDKNWATLAQKLGLGILNNAFRLSPAPSKTLMDNYEVSGGTVRELVEALRQMGYTEAIEVIQAASSPVKTTSQAHSLPLSPASTRQQIDELRDSDSVCDSGVETSFRKLSFTESLTSGASLLTLNKMPHDYGQEGPLEGKI\n",
      "105338.04820000062 g/mol\n",
      "60201.2142100002 g/mol\n"
     ]
    }
   ],
   "source": [
    "seq = open('P19838.fasta.txt')\n",
    "seq_NFKBI = []\n",
    "for line in seq:\n",
    "    if line.startswith('>'):\n",
    "        continue\n",
    "    else:\n",
    "        seq_NFKBI.append(line.strip())\n",
    "seq_NFKBI = ''.join(seq_NFKBI)\n",
    "print(seq_NFKBI)\n",
    "mass_NFKBI = mass(seq_NFKBI)\n",
    "print(mass_NFKBI, 'g/mol')\n",
    "\n",
    "seq2 = open('Q04206.fasta.txt')\n",
    "seq_RELA = []\n",
    "for line in seq2:\n",
    "    if line.startswith('>'):\n",
    "        continue\n",
    "    else:\n",
    "        seq_RELA.append(line.strip())\n",
    "seq_RELA = ''.join(seq_RELA)\n",
    "mass_RELA = mass(seq_RELA)\n",
    "print(mass_RELA, 'g/mol')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.7 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "1c0e22491d37f29e12dbbed1764984b5047c336fcb37effe27c9c012a2ac9d2b"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
